openlibrary/core/bookshelves.py

```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
         result = oldb.query(query, vars=data)
         return {i['bookshelf_id']: i['count'] for i in result} if result else {}
 
+    # Iterates through a list of solr docs, and for all items with a 'logged edition'
+    # it will remove an item with the matching edition key from the list, and add it to
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
+        linked_docs: list[web.storage] = []
+        docs_dict = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            full_key = doc.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "works":
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    docs_dict.update(
+                        {doc["logged_edition"].split("/")[2]: (len(linked_docs) - 1)}
```

Comment:
We should be able to have this map directly to the doc to make it a bit easier to read!

```suggestion
                        {doc["logged_edition"]: doc}
```

Reply:
As mentioned before, it's nice to remember that python makes such heavy use of references.  My previous version mapped keys to indices, but I'm now mapping keys to docs, as in this request.